File: go/core/registry.go
package core

import (
	"fmt"
	"strings"
)

// Registry — Trung tâm đăng ký module.
// Upgraded: Tích hợp EventBus, Middleware pipeline, Lifecycle hooks.
type Registry struct {
	modules     map[string]Module
	middlewares []Middleware
	EventBus    *EventBus
}

// NewRegistry khởi tạo registry rỗng.
func NewRegistry() *Registry {
	return &Registry{
		modules:  make(map[string]Module),
		EventBus: NewEventBus(),
	}
}

// Register đăng ký 1 module. Trùng tên → panic.
func (r *Registry) Register(m Module) {
	name := strings.ToLower(m.Name())
	if _, exists := r.modules[name]; exists {
		panic(fmt.Sprintf("module '%s' already registered", name))
	}
	r.modules[name] = m
}

// RegisterWithLifecycle đăng ký module với lifecycle hooks.
func (r *Registry) RegisterWithLifecycle(m Module) error {
	if lc, ok := m.(ModuleLifecycle); ok {
		if err := lc.OnInitializing(); err != nil {
			return fmt.Errorf("OnInitializing failed for '%s': %w", m.Name(), err)
		}
	}

	r.Register(m)

	if lc, ok := m.(ModuleLifecycle); ok {
		if err := lc.OnInitialized(); err != nil {
			return fmt.Errorf("OnInitialized failed for '%s': %w", m.Name(), err)
		}
	}
	return nil
}

// AddMiddleware thêm middleware vào pipeline. FIFO order.
func (r *Registry) AddMiddleware(mw Middleware) {
	r.middlewares = append(r.middlewares, mw)
}

// Dispatch chuyển command tới module, qua middleware pipeline.
// Pipeline executes module at end of chain — middleware can read ctx.Result after next().
func (r *Registry) Dispatch(moduleName, command string, args []string) string {
	m, ok := r.modules[strings.ToLower(moduleName)]
	if !ok {
		keys := make([]string, 0, len(r.modules))
		for k := range r.modules {
			keys = append(keys, k)
		}
		return fmt.Sprintf("module '%s' not found. Available: %s", moduleName, strings.Join(keys, ", "))
	}

	if len(r.middlewares) == 0 {
		return m.Execute(command, args)
	}

	ctx := NewContext(moduleName, command, args)
	r.runPipeline(ctx, m)
	return ctx.Result
}

func (r *Registry) runPipeline(ctx *ModuleContext, m Module) {
	index := -1
	var next func()
	next = func() {
		index++
		if index < len(r.middlewares) {
			r.middlewares[index].Invoke(ctx, next)
		} else {
			// End of pipeline → execute module. ctx.Result available for middleware after next().
			ctx.Result = m.Execute(ctx.Command, ctx.Args)
		}
	}
	next()
}

// Shutdown tất cả modules có lifecycle.
func (r *Registry) Shutdown() error {
	for _, m := range r.modules {
		if lc, ok := m.(ModuleLifecycle); ok {
			if err := lc.OnShuttingDown(); err != nil {
				return err
			}
			if err := lc.OnShutdown(); err != nil {
				return err
			}
		}
	}
	return nil
}

// GetAll trả về tất cả modules.
func (r *Registry) GetAll() map[string]Module {
	return r.modules
}

// Count trả về số lượng modules.
func (r *Registry) Count() int {
	return len(r.modules)
}

// MiddlewareCount trả về số middleware.
func (r *Registry) MiddlewareCount() int {
	return len(r.middlewares)
}

File: python/core/registry.py
"""
Registry — Trung tâm đăng ký module.
Upgraded: EventBus, Middleware pipeline, Lifecycle hooks.
"""
import asyncio
from .module import IModule
from .eventbus import EventBus
from .lifecycle import IModuleLifecycle
from .middleware import IMiddleware, ModuleContext


class ModuleRegistry:
    """Registry quản lý và dispatch commands tới modules qua middleware pipeline."""

    def __init__(self) -> None:
        self._modules: dict[str, IModule] = {}
        self._middlewares: list[IMiddleware] = []
        self.event_bus = EventBus()

    def register(self, module: IModule) -> None:
        """Đăng ký 1 module. Trùng tên → raise."""
        key = module.name.lower()
        if key in self._modules:
            raise ValueError(f"Module '{module.name}' already registered.")
        self._modules[key] = module

    async def register_async(self, module: IModule) -> None:
        """Đăng ký module với lifecycle hooks."""
        if isinstance(module, IModuleLifecycle):
            await module.on_initializing()
        self.register(module)
        if isinstance(module, IModuleLifecycle):
            await module.on_initialized()

    def add_middleware(self, mw: IMiddleware) -> None:
        """Thêm middleware vào pipeline. FIFO."""
        self._middlewares.append(mw)

    def dispatch(self, module_name: str, command: str, args: list[str]) -> str:
        """Dispatch command qua middleware pipeline (sync)."""
        module = self._modules.get(module_name.lower())
        if module is None:
            available = ", ".join(self._modules.keys())
            raise KeyError(f"Module '{module_name}' not found. Available: {available}")

        if not self._middlewares:
            return module.execute(command, args)

        context = ModuleContext(
            module_name=module_name,
            command=command,
            args=args,
        )
        self._run_pipeline(context, module)
        return context.result if context.result is not None else module.execute(command, args)

    async def dispatch_async(self, module_name: str, command: str, args: list[str]) -> str:
        """Dispatch command (async). Cho phép module hỗ trợ async execution."""
        module = self._modules.get(module_name.lower())
        if module is None:
            available = ", ".join(self._modules.keys())
            raise KeyError(f"Module '{module_name}' not found. Available: {available}")

        if not self._middlewares:
            return module.execute(command, args)

        context = ModuleContext(
            module_name=module_name,
            command=command,
            args=args,
        )
        self._run_pipeline(context, module)
        return context.result if context.result is not None else module.execute(command, args)

    def _run_pipeline(self, context: ModuleContext, module: IModule) -> None:
        index = [-1]  # mutable counter

        def next_fn() -> None:
            index[0] += 1
            if index[0] < len(self._middlewares):
                self._middlewares[index[0]].invoke(context, next_fn)
            else:
                context.result = module.execute(context.command, context.args)

        next_fn()

    async def shutdown(self) -> None:
        """Shutdown tất cả modules có lifecycle."""
        for module in self._modules.values():
            if isinstance(module, IModuleLifecycle):
                await module.on_shutting_down()
                await module.on_shutdown()

    def get_all(self) -> dict[str, IModule]:
        return dict(self._modules)

    @property
    def count(self) -> int:
        return len(self._modules)

    @property
    def middleware_count(self) -> int:
        return len(self._middlewares)
